tool/tsh/resolve_default_addr.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+type raceResult struct {
+	addr string
+	err  error
+}
+
+// raceRequest drives an HTTP request to completion and posts the results back
+// to the supplied channel.
+func raceRequest(ctx context.Context, cli *http.Client, addr string, results chan<- raceResult) {
+	target := fmt.Sprintf("https://%s/", addr)
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, target, nil)
+
+	if err == nil {
+		var rsp *http.Response
+		rsp, err = cli.Do(request)
```

Comment:
What if the request succeeds but returns non-200 response?

Reply:
For the purposes of determining which a port a server might accessible on, I'd argue that a non-OK response is probably as good as an OK one.

Reply:
Are you sure this is an assumption we can safely make? I.e. what if the service is behind a load balancer or a reverse proxy that returns 500 in case its backend is down? Would there be any downside to making sure this returns a 200 response?

Reply:
Done